src/Stack/Config.hs

```diff
@@ -611,15 +627,78 @@ resolvePackageLocation menv projRoot (PLRemote url remotePackageType) = do
-determineStackRoot :: (MonadIO m, MonadThrow m) => m (Path Abs Dir)
-determineStackRoot = do
-    env <- liftIO getEnvironment
-    case lookup stackRootEnvVar env of
-        Nothing -> getAppUserDataDir $(mkRelDir stackProgName)
-        Just x -> do
-            liftIO $ D.createDirectoryIfMissing True x
-            resolveDir' x
+-- | Get the stack root, e.g. @~/.stack@, and determine whether the user owns it.
+--
+-- On Windows, the second value is always 'True'.
+determineStackRootAndOwnership
+    :: (MonadIO m, MonadCatch m)
+    => m (Path Abs Dir, Bool)
+determineStackRootAndOwnership = do
+    stackRoot <- do
+        mstackRoot <- liftIO $ lookupEnv stackRootEnvVar
+        case mstackRoot of
+            Nothing -> getAppUserDataDir $(mkRelDir stackProgName)
+            Just x -> parseAbsDir x
+
+    (existingStackRootOrParentDir, userOwnsIt) <- do
+        mdirAndOwnership <- findInParents getDirAndOwnership stackRoot
+        case mdirAndOwnership of
+            Just x -> return x
+            Nothing -> throwM (BadStackRootEnvVar stackRoot)
+
+    when (existingStackRootOrParentDir /= stackRoot) $
+        if userOwnsIt
+            then liftIO $ ensureDir stackRoot
+            else throwM $
+                Won'tCreateStackRootInDirectoryOwnedByDifferentUser
+                    stackRoot
+                    existingStackRootOrParentDir
+
+    stackRoot' <- canonicalizePath stackRoot
+    return (stackRoot', userOwnsIt)
+
+-- | @'checkOwnership' dirs@ throws 'UserDoesn'tOwnDirectory' if the first
+-- existing directory of @dirs@ isn't owned by the current user.
+--
+-- If none of the directories exist, throws @'NoSuchDirectory' lastDir@, where
+-- @lastDir@ is @last dirs@.
+checkOwnership :: (MonadIO m, MonadCatch m) => NonEmpty (Path Abs Dir) -> m ()
+checkOwnership dirs = do
+    mdirAndOwnership <- firstJustM getDirAndOwnership (NE.toList dirs)
+    case mdirAndOwnership of
+        Just (_, True) -> return ()
+        Just (dir, False) -> throwM (UserDoesn'tOwnDirectory dir)
+        Nothing ->
+            (throwM . NoSuchDirectory . toFilePathNoTrailingSep . NE.last) dirs
+
+-- | @'getDirAndOwnership' dir@ returns @'Just' (dir, 'True')@ when @dir@
+-- exists and the current user owns it in the sense of 'isOwnedByUser'.
+getDirAndOwnership
+    :: (MonadIO m, MonadCatch m)
+    => Path Abs Dir
+    -> m (Maybe (Path Abs Dir, Bool))
+getDirAndOwnership dir = forgivingAbsence $ do
+    ownership <- isOwnedByUser dir
+    return (dir, ownership)
+
+-- | Check whether the current user (determined with 'getEffectiveUserId') is
+-- the owner for the given path.
+--
+-- Will always return 'True' on Windows.
+isOwnedByUser :: MonadIO m => Path Abs t -> m Bool
+isOwnedByUser path = liftIO $ do
+    if osIsWindows
+        then return True
+        else do
+            fileStatus <- getFileStatus (toFilePath path)
+            user <- getEffectiveUserID
+            return (user == fileOwner fileStatus)
```

Comment:
On a second look, I wonder if this couldn't cause some hard to find Windows-only bugs.

I guess we should better get the file status on Windows, too, so it will cause an exception when the file or directory doesn't exist.